### PR TITLE
Boos 453/feature/redis celery

### PR DESCRIPTION
**Pull request by koutsilisk**
In this PR we introduce:
- Celery setup.
- Celery docker container.
- Redis docker container.

### CODE REVIEW

File: backend/.gitignore [hunk 1]
```diff
@@ -3,5 +3,6 @@
 __pycache__ 
 .vscode
 .cache/
+*.pyc
```
Comment (Chichilele) [inline backend/.gitignore:6]: oh good spot!

File: backend/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,7 @@
 FROM python:3.8
 
+ENV LANG=C.UTF-8 LC_ALL=C.UTF-8 PYTHONUNBUFFERED=1
```
Comment (Chichilele) [inline backend/Dockerfile:3]: can you remind me what was this for? Was it celery or general python stuff?

File: backend/requirements.txt [hunk 3]
```diff
@@ -18,16 +25,18 @@ importlib-resources==5.4.0
...
 Jinja2==3.0.2
+kombu==5.2.3
 Mako==1.1.5
 MarkupSafe==2.0.1
 marshmallow==3.14.0
 marshmallow-sqlalchemy==0.26.1
 mkdocs-material-extensions==1.0.3
 mypy-extensions==0.4.3
-packaging==21.2
+packaging==21.3
```
Comment (Chichilele) [inline backend/requirements.txt:35]: careful, we're upgrading an existing dependency (if everything works I suppose it's fine but let's add it to all PRs)

File: docker-compose.yml [hunk 4]
```diff
@@ -18,6 +18,20 @@ services:
       timeout: 5s
       retries: 3
   
+  redis:
```
Comment (Chichilele) [inline docker-compose.yml:21]: Maybe we can find a high level naming... I'm sorry I can't think of anything right now but to keep in mind. I would have said `broker` but it's more than that... `ram_db`? idk

File: docker-compose.yml [hunk 5]
```diff
@@ -34,12 +48,33 @@ services:
     depends_on:
       postgres:
         condition: service_healthy
+      redis:
```
Comment (Chichilele) [inline docker-compose.yml:51]: Let's also watch how many dependencies we add. We may end up with a very sequential startup which means it'll be slow. Some services may require others to work but can be started before their dependencies are READY

File: backend/project/tasks.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from project.app import celery
+
+
+@celery.task
```
Comment (Chichilele) [inline backend/project/tasks.py:4]: How about the test setup for celery functions?

File: backend/project/app.py [hunk 7]
```diff
@@ -20,6 +21,9 @@
 cors = CORS()
 mail = Mail()
 health = HealthCheck()
+celery = Celery(
```
Comment (Chichilele) [inline backend/project/app.py:24]: Celery should be in its own module, implementing `XXX.init_app(app, ARGS)`

**Review by Chichilele — CHANGES_REQUESTED** [review]
create a celery extension

File: backend/Dockerfile (hunk 2 above)
Comment (koutsilisk) [reply backend/Dockerfile:3]: This is python stuff, the pythonunbuffered is for logs. I saw them in other dockerfiles so I added them too.

File: backend/requirements.txt (hunk 3 above)
Comment (koutsilisk) [reply backend/requirements.txt:35]: this got upgraded when installing celery.

File: docker-compose.yml (hunk 5 above)
Comment (koutsilisk) [reply docker-compose.yml:51]: good point.

File: backend/project/tasks.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-from project.app import celery
+from project.extensions.celery import celery
```
Comment (Chichilele) [inline backend/project/tasks.py:1]:
```python
from project.extensions import celery
```

File: backend/project/extensions/celery.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from celery import Celery
+from config import BaseConfig
+
```
Comment (Chichilele) [inline backend/project/extensions/celery.py:3]: Implement extension interface with init_ap() method

**Review by Chichilele — CHANGES_REQUESTED** [review]

File: backend/project/extensions/celery.py (hunk 9 above)
Comment (Chichilele) [reply backend/project/extensions/celery.py:3]: done

File: backend/project/tasks.py (hunk 8 above)
Comment (Chichilele) [reply backend/project/tasks.py]: fixed

**Review by Chichilele — CHANGES_REQUESTED** [review]
- [ ] Since you added a dummy task, add a dummy test
- Are you sure the code you borrowed is fine?

File: backend/project/extensions/celery.py [hunk 10]
```diff
@@ -1,13 +1,17 @@
...
+class FlaskCelery(Celery):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
+        if "app" in kwargs:
+            self.init_app(kwargs["app"])
+
     def init_app(self, app):
         self.conf.update(app.config)
+        self.autodiscover_tasks("project")
```
[the remark this answers is not in the document]
Comment (koutsilisk) [reply backend/project/extensions/celery.py:14]: the if app in kwargs is in order to be able to init with an app or without and app, ``` celery = FlaskCelery(app) or celery = FlaskCelery() ``` the autodiscover_tasks i saw it from biran capello but checked what it does.,basically it tries to find a task module under the project folder ( in that case )

**Comment by Chichilele** [comment]
Closing for rebase

**Pull request closed without merging**